### PR TITLE
JavaScript: Make `File` not extend `Locatable` anymore.

### PR DESCRIPTION
**Pull request by max-schaefer**
Files have strange `:0:0:0:0` locations for... reasons. This makes the predicates inherited from `Locatable` meaningless. A particularly bad case is `getNumLines()`, which will always return 1. The right predicate to use is, of course, `getNumberOfLines()`, which is defined in `File` itself :trollface: 

I have verified that none of our existing queries rely on `File` being a subtype of `Locatable`. Note that I explicitly did _not_ change the dbscheme type hierarchy, where `@file` is still a subtype of `@locatable`. Hence `File` is still compatible with `Locatable`, you just can't invoked the latter's member predicates without an explicit cast.

The second commit is slightly orthogonal, but without it a query that selects `File`s as `Locatable`s will fail to display its results properly, since it will try to use the `getLocation()` predicate in `Locatable`, which previously was implemented as `none()`, relying on subclasses to fill in the correct implementation. Since that's almost always `hasLocation(this, result)` we might as well use _that_ as the default implementation, reducing the chance of surprising behaviour involving `File`s.

### CODE REVIEW

**Comment by max-schaefer** [comment]
Marking as WIP for now: even if we decide we want to do this, we need to check that no customer queries are affected before merging. @esbena, @asgerf, @rdmarsh2: let's coordinate on this.

I'll add a change note once we are all happy.

File: javascript/ql/src/semmle/javascript/Files.qll [hunk 1]
```diff
@@ -198,8 +198,8 @@ class Folder extends Container, @folder {
 }
 
 /** A file. */
-class File extends Container, @file, Locatable {
-  override Location getLocation() { hasLocation(this, result) }
+class File extends Container, @file {
+  Location getLocation() { hasLocation(this, result) }
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/Files.qll]: Needs qldoc now

**Comment by max-schaefer** [comment]
> Do you want to run an evaluation for this?

Might as well; better safe than sorry.

**Comment by esbena** [comment]
None of the custom queries at my customers use:

- `Locatable`
- `getEndLine`
- `getFile` (lol)
- `getLocation`
- `getNumLines`
- `getNumLines`
- `getStartLine`

In relation to `File`s.

**Comment by max-schaefer** [comment]
I've had a bit of a fun time with doing a performance evaluation for this.

My original proposal caused subtle changes to dispatch for `getLocation`, which caused performance of the range analysis library (!) to regress, due to worse join-ordering of a predicate doing location comparison. I fixed that, but it turns out that there were a few other smaller regressions that together added up to something non-negligible.

So I decided to abandon that idea, and instead went with a somewhat sneaky private subclass of `File` and `Locatable` that patches up the lost dynamic dispatch in a different way. That seems to be mostly [performance neutral](https://git.semmle.com/max/dist-compare-reports/blob/master/js/file-locatable/report.md) (happy to re-run anything you'd like me to).

To round things off, the tests also exposed one place where we really did rely on `File` being `Locatable`, in the dependencies library. This was easy to fix, though, and the API isn't used frequently enough to make much of a fuss about it.

Ready for re-review, please take a look.

**Pull request merged**